tools/code-generation/generator/src/main/java/com/amazonaws/util/awsclientgenerator/domainmodels/codegeneration/Shape.java

```diff
@@ -219,23 +220,49 @@ public String toString() {
       return "null";
     }
 
+    private void getAllChildrenShapes(Map<String, Shape> collected) {
+        if (members == null || members.isEmpty()) {
+            return;
+        }
+
+        Map<String, Shape> thisChildren = members.values().parallelStream()
+                .map(value -> value.getShape())
+                .filter(shape -> !collected.containsKey(shape.getName()))
+                .map(memberShape -> memberShape.isList() ? memberShape.getListMember().getShape() : memberShape) //if references through a list container
+                .collect(Collectors.toMap(
+                        shape -> shape.getName(), shape -> shape,
+                        (shape1, shape2) -> shape1));
+
+        collected.putAll(thisChildren);
+
+        for(Shape child : thisChildren.values()) {
+            child.getAllChildrenShapes(collected);
+        }
+    }
+
+    private Set<String> thisAndAllChildrenShapeNamesMemoized;
+    private Set<String> getThisAndAllChildrenShapeNames() {
+        if (thisAndAllChildrenShapeNamesMemoized != null && !thisAndAllChildrenShapeNamesMemoized.isEmpty()) {
+            // this method is called for at least shape count ^ 2 times
+            return thisAndAllChildrenShapeNamesMemoized;
+        }
+        Map<String, Shape> shapes = new HashMap<String, Shape>();
+
+        getAllChildrenShapes(shapes);
+
+        thisAndAllChildrenShapeNamesMemoized = new HashSet<>(shapes.keySet());
```

Comment:
nit: instead of creating then assigning it, we can just make it a instance of immutable set

```suggestion
      thisAndAllChildrenShapeNamesMemoized = ImmutableSet.<String>builder()
                .addAll(shapes.keySet()).
                add(getName())
                .build();
        return thisAndAllChildrenShapeNamesMemoized;
```